nat-lab/tests/teliod.py

```diff
@@ -0,0 +1,310 @@
+        self,
+        connection: Connection,
+        exit_stack: AsyncExitStack,
+        config: Config = Config(),
+    ) -> None:
+        self._connection: Connection = connection
+        self._exit_stack: AsyncExitStack = exit_stack
+        self.config: Config = config
+
+    async def execute_command(
+        self,
+        cmd: Command,
+    ) -> tuple[str, str]:
+        try:
+            proc = await self._connection.create_process(cmd).execute()
+            stdout, stderr = proc.get_stdout(), proc.get_stderr()
+            log.debug("'%s' stdout: '%s', stderr: '%s'", cmd, stdout, stderr)
+            return stdout, stderr
+        except ProcessExecError as exc:
+            log.debug("Exception occured while executing teliod command: %s", exc)
+            raise
+
+    async def run_command(
+        self,
+        cmd: Command,
+    ) -> "Process":
+        proc = await self._exit_stack.enter_async_context(
+            self._connection.create_process(cmd).run()
+        )
+        return proc
+
+    @asynccontextmanager
+    async def start(self) -> AsyncIterator["Teliod"]:
+        try:
+            await self.remove_logs()
+
+            async def wait_for_teliod_start():
+                await self.wait_for_teliod_socket()
+                while True:
+                    try:
+                        if not await self.is_alive():
+                            raise RuntimeError(
+                                "socket exists but daemon's not running."
+                            )
+                        break
+                    except TeliodObtainingIdentity:
+                        await asyncio.sleep(self.TELIOD_CMD_CHECK_INTERVAL_S)
+                        continue
+
+            if not self.config.no_detach:
+                stdout, stderr = await self.execute_command(Command.start(self.config))
+                await wait_for_teliod_start()
+            else:
+                proc = await self.run_command(Command.start(self.config))
+                await wait_for_teliod_start()
+                stdout, stderr = proc.get_stdout(), proc.get_stderr()
+
+            assert len(stderr) == 0, f"Stderr is not empty: {stderr}"
+            await self.config.assert_match_daemon_start(stdout)
+            yield self
+        finally:
+            try:
+                await self.quit()
+            except ProcessExecError as exc:
+                if "Error: DaemonIsNotRunning" not in exc.stderr:
+                    log.error(exc)
+                    await self.kill()
+                    await self.remove_socket()
+                else:
+                    log.info("Tried to quit but daemon is already not running")
+
+    async def is_alive(self) -> bool:
+        try:
+            stdout, _ = await self.execute_command(Command.is_alive())
+            return "Command executed successfully" in stdout
+        except ProcessExecError as exc:
+            if "Obtaining identity, ignoring" in exc.stdout:
```

Comment:
This might need to be
```suggestion
            if "Obtaining identity, ignoring" in exc.stderr:
```

Reply:
Thanks, how did you check it?